examples/src/main/java/io/grpc/examples/helloworld/HelloWorldClient.java

```diff
@@ -0,0 +1,268 @@
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.examples.helloworld;
+
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.util.RoundRobinLoadBalancerFactory;
+import io.grpc.StatusRuntimeException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import java.net.URI;
+import java.util.Comparator;
+import java.util.List;
+import java.util.ArrayList;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import javax.annotation.Nullable;
+
+import io.grpc.Attributes;
+import io.grpc.NameResolver;
+import io.grpc.NameResolverProvider;
+import io.grpc.EquivalentAddressGroup;
+//
+// import zookeeper classes
+import java.util.concurrent.CountDownLatch;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.AsyncCallback.StatCallback;
+import org.apache.zookeeper.KeeperException.Code;
+import org.apache.zookeeper.data.Stat;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.ZooDefs;
+import org.apache.zookeeper.data.Stat;
+
+
+class ZkNameResolver extends NameResolver implements Watcher {
+  /* Hard-coded path to the ZkNode that knows about servers */
+  public static final String path = "/grpc_hello_world_service";
+  /* 2 seconds to detect that server disconnected */
+  public static final int kTimeoutMs = 2000;
+
+  private static URI zk_uri;
+  private static ZooKeeper zoo;
+  private static Listener listener;
+  private static final Logger logger = Logger.getLogger("ZK");
+
+  /*
+   * The callback from Zookeeper when servers are added/removed.
+   */
+  public void process(WatchedEvent we) {
+    if (we.getType() == Event.EventType.None) {
+      logger.info("Connection expired");
+    } else {
+      try {
+        List<String> servers  = zoo.getChildren(path, false);
+        AddServersToListener(servers);
+        zoo.getChildren(path, this);
```

Comment:
The first call to getChildren is synchronous and the second one is async. They have to be separate, as the second one will be invoked when server status changes.